lib/logstash/config/config_ast.rb

```diff
@@ -309,11 +319,36 @@ def compile
     end
   end
   class Hash < Value
+    def validate!
+      duplicate_values = find_duplicate_keys
+
+      if duplicate_values.size > 0
+        raise ConfigurationError.new(
+          I18n.t("logstash.agent.configuration.invalid_plugin_settings_duplicated_keys",
+                 :keys => duplicate_values.join(', '),
+                 :line => input.line_of(interval.first),
+                 :column => input.column_of(interval.first),
+                 :byte => interval.first + 1,
+                 :after => input[0..interval.first]
+                )
+        )
+      end
+    end
+
+    def find_duplicate_keys
+      values = recursive_select(HashEntry).collect { |hash_entry| hash_entry.name.text_value_utf8 }
+      values.find_all { |v| values.count(v) > 1 }.uniq
+    end
+
     def compile
-      return "{" << recursive_select(HashEntry).collect(&:compile).reject(&:empty?).join(", ") << "}"
+      validate!
+      return "{" << recursive_select(HashEntry).compact.collect(&:compile).reject(&:empty?).join(", ") << "}"
```

Comment:
I think it's not needed, since I removed the changes in the grammar. I will remove if it  doesnt break anything.